File: rust/src/cpf.rs
use lazy_static::lazy_static;
use pad::{PadStr, Alignment};
use regex::Regex;
use std::vec::IntoIter;
use substring::Substring;


pub fn verificar(cpf: &str) -> bool {
    /*
    - cpf: string (texto) com o número de CPF, com ou sem marcação; o
        número será tratado como um CPF completo; seus dígitos finais
        serão comparados com os dígitos calculados a partir dos 9
        primeiros dígitos.
    */
    if reter_numeros(cpf, 1) == "" {
         println!("ERRO: informe um CPF válido para verificação");
         return false
    }
    let mut digitos_recebidos = vec![];
    let mut c: u8 = 0;
    for d in obter_digitos(cpf, 11) {
        if c == 9 || c == 10 {
            digitos_recebidos.extend(vec![d]);
        }
        c += 1
    }
    let cpf_calculo = reter_numeros(cpf, 11).to_string();
    let digitos_calculados: [u16; 2] = calcular_digitos(&cpf_calculo.substring(0, 9));
    if digitos_recebidos[0] == digitos_calculados[0]
        && digitos_recebidos[1] == digitos_calculados[1] {
        return true
    }
    false
}

pub fn obter_digitos(cpf: &str, n: usize) -> IntoIter<u16> {
    /*
    - cpf: string com números que serão filtrados e transformados em
        inteiros para execução do cálculo de CPF;
    - n: inteiro que corresponde à quantidade de dígitos desejada, a
        depender da operação em questão:
        -  9 dígitos -> Cálculo dos dígitos verificadores de CPF;
        - 11 dígitos -> Verificação do número de CPF.
    */
    let cpf_filtrado = &reter_numeros(&cpf, n);
    let mut digitos = vec![];
    let mut i: usize = 0;
    while i < cpf_filtrado.len() && i < n {
        let d = cpf_filtrado.chars().nth(i).unwrap().to_string().parse::<u16>().unwrap();
        digitos.extend(vec![d]);
        i += 1
    }
    digitos.into_iter()
}

pub fn calcular_digitos(cpf: &str) -> [u16; 2] {
    /*
    - cpf: string (texto) com o número de CPF, com ou sem marcação;
        o número será encarado como um CPF incompleto (sem os dígitos
        verificadores); seus dígitos finais serão calculados pela
        função 'calcular_digito_verificador'.
    */
    if reter_numeros(cpf, 9) == "" {
        panic!("ERRO: informe um número válido.")
    }
    let mut digitos_cpf = vec![];
    for d in obter_digitos(cpf, 9) {
        digitos_cpf.extend(vec![d]);
    }
    let digito_verificador_1 = calcular_digito_verificador(&digitos_cpf);
    digitos_cpf.extend(vec![digito_verificador_1]);
    [ digito_verificador_1, calcular_digito_verificador(&digitos_cpf) ]
}

pub fn reter_numeros(cpf: &str, n: usize) -> String {
    /*
    - cpf: ponteiro para string com números que serão isolados
        de caracteres não numéricos, e então preenchidos com zeros,
        para que a string tenha o mesmo comprimento que 'n'; caso
        não existam caracteres numéricos, será devolvida uma
        string vazia.
    - n: inteiro que corresponde à quantidade de dígitos desejada (e
        também ao comprimento final da string retornada, a depender
        da operação em questão:
        -  9 dígitos -> Cálculo dos dígitos verificadores de CPF;
        - 11 dígitos -> Verificação do número de CPF.
    */
    lazy_static! {
        static ref REGEXP_NUMBERS: Regex = Regex::new(r"\d").unwrap();
    }
    lazy_static! {
        static ref REGEXP_NON_NUMBERS: Regex = Regex::new(r"\D").unwrap();
    }
    if REGEXP_NUMBERS.is_match(cpf) {
        let number_string = REGEXP_NON_NUMBERS.replace_all(cpf, "").to_string();
        return number_string.pad(n, '0', Alignment::Right, true)
    }
    String::from("")
}

fn calcular_digito_verificador(digitos: &[u16]) -> u16 {
    /*
    - digitos: slice contendo 9 ou 10 dígitos para efetuar o cálculo
        dos dígitos verificadores.
    */
    let mut soma: u16 = 0;
    let mut multiplicador: u16 = u16::try_from(digitos.len()).unwrap();
    multiplicador += 1;
    for d in digitos {
        soma += d * multiplicador;
        multiplicador -= 1
    }
    let resto: u16 = soma % 11;
    if resto > 1 {
        return 11 - resto
    }
    0
}

File: lua/cpf.lua
-- LuaJIT 2.0.5 -- Copyright (C) 2005-2017 Mike Pall. http://luajit.org/
local params = {...}

function Verificar(cpf)
    local c = ReterNumeros(cpf, 11)
    if string.len(c) < 1 then
        print("ERRO: o CPF deve conter pelo menos um caracter numérico.")
        return end
    local Cpf = string.sub(c, 1, 3)
        .."."..string.sub(c, 4, 6)
        .."."..string.sub(c, 7, 9)
        .."-"..string.sub(c, 10, 11)
    print("               CPF:", Cpf)
    local digitos = ObterDigitos(cpf, 11)
    local dvCpf1, dvCpf2 = digitos[10], digitos[11]
    print("       Dígitos CPF:", dvCpf1, dvCpf2)
    c = string.sub(ReterNumeros(cpf, 11), 1, 9)
    local dvCalc1, dvCalc2 = CalcularDigitos(c)
    print("Dígitos calculados:", dvCalc1, dvCalc2)
    if dvCpf1 == dvCalc1 and dvCpf2 == dvCalc2 then
        return true end
    return false end

function ObterDigitos(cpf, n)
    local digitos = {}
    local Cpf = ReterNumeros(cpf, n)
    if Cpf == "" then
        return digitos end
    for i=1, n, 1 do
        local d = ParseInt(string.sub(Cpf, i, i))
        table.insert(digitos, d) end
    return digitos end

function ReterNumeros(cpf, n)
    local Cpf = string.gsub(cpf, "%D", "")
    if Cpf == "" then
        return Cpf end
    while string.len(Cpf) < n do
        Cpf = string.rep("0", n - string.len(Cpf))..Cpf end
    return Cpf end

function CalcularDigitos(cpf)
    if string.len(ReterNumeros(cpf, 1)) < 1 then
        print("ERRO: o CPF deve conter pelo menos um caracter numérico.")
        return end
    local digitos = ObterDigitos(cpf, 9)
    local dv1 = CalcularDigito(digitos)
    table.insert(digitos, dv1)
    local dv2 = CalcularDigito(digitos)
    local Cpf = ReterNumeros(cpf, 9)
    local dvs = string.format("%d%d", dv1, dv2)
    print(string.sub(Cpf, 1, 3).."."..string.sub(Cpf, 4, 6)
        .."."..string.sub(Cpf, 7, 9).."-"..dvs)
    print(Cpf..dvs)
    return dv1, dv2 end

function ParseInt(str)
    function hasInteger(str)
        return not (str == "" or str:match("%D")) end
    function hasFloat(str)
        return not not (str:match("%d.%d")) end
    if hasInteger(str) or hasFloat(str) then
        return tonumber(str) end
    return false end

function CalcularDigito(digitos)
    local multiplicador = table.getn(digitos) + 1
    local soma = 0
    for i=1, table.getn(digitos), 1 do
        soma = soma + (multiplicador * digitos[i])
        multiplicador = multiplicador - 1 end
    local resto = soma % 11
    if resto > 1 then
        return 11 - resto end
    return 0 end

if params[1] == '-c' or params[1] == '--calcular' then
    print(CalcularDigitos(params[2])) end
if params[1] == '-v' or params[1] == '--verificar' then
    print(Verificar(params[2])) end
if params[1] == '--demo' then
    -- CPF inválido p/ verificação (sem números)
    local a = "a"
    print('1) CPF = "'..a..'"')
    print('   - Verificar("'..a..'")')
    print(Verificar(a)) -- expected: false
    print('\n   - CalcularDigitos("'..a..'")')
    print(CalcularDigitos(a))
   -- CPF válido p/ verificacão (o número é inválido)
   print("\n")
    local b = "123a"
    print('2) CPF = "'..b..'"')
    print('   - Verificar("'..b..'")')
    print(Verificar(b)) -- expected: false
    print('\n   - CalcularDigitos("'..b..'")')
    print(CalcularDigitos(b))
    -- CPF válido p/ verificação (o número é válido)
    print("\n")
    local c = "123-60"
    print('3) CPF = "'..c..'"')
    print('   - Verificar("'..c..'")')
    print(Verificar(c)) -- expected: true
    print('\n   - CalcularDigitos("'..c..'")')
    print(CalcularDigitos(c)) end
